e2e: wait for pods with gomega

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

gomega.Eventually provides better progress reports: instead of filling up the
log with one-line messages that are not enough to to understand the
current state, it integrates with Gingko's progress reporting (SIGUSR1,
--poll-progress-after) and then dumps the same complete failure message as
after a timeout. That makes it possible to understand why progress isn't
getting made without having to wait for the timeout.
    
The code also is generic enough that it becomes more reusable:
- error handling (= retry after "not found", delay) can be used
  for all kind of objects
- multiple checks can be combined with gomega.And

#### Which issue(s) this PR fixes:
Related-to #https://github.com/kubernetes/kubernetes/issues/106575

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
e2e framework: implement pod polling with gomega.Eventually

gomega.Eventually provides better progress reports: instead of filling up the
log with rather useless one-line messages that are not enough to to understand
the current state, it integrates with Gingko's progress reporting (SIGUSR1,
--poll-progress-after) and then dumps the same complete failure message as
after a timeout. That makes it possible to understand why progress isn't
getting made without having to wait for the timeout.

The other advantage is that the failure message for some unexpected pod state
becomes more readable: instead of encapsulating it as "observed object" inside
an error, it directly gets rendered by gomega.

pohly commented:
/sig testing

aojea commented on test/e2e/framework/wait.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+// GetFunc is a function which retrieves a certain object.
+type GetFunc[T any] func(ctx context.Context) (T, error)
```
heh, so this is the first Generic in the codebase? 😄

pohly replied:
It's still in the race, but https://github.com/kubernetes/kubernetes/pull/112377 will probably be faster :grin:

leilajal commented:
/remove-sig api-machinery

dims commented:
If you still need this PR then please rebase, if not, please close the PR

dims commented:
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

pohly commented:
I'll get back to it soonish after the other E2E enhancements have landed.

pohly commented:
There are some more tests left which (incorrectly) use `framework.ExpectError` and wait for pods to be running which are not expected to start. I'll fix this next week.

pohly committed:
e2e framework: gomega assertions as errors

Calling gomega.Expect/Eventually/Consistently deep inside a helper call chain
has several challenges:
- the stack offset must be tracked correctly, otherwise the callstack
  for the failure starts at some helper code, which is often not informative
- augmenting the failure message with additional information from each
  caller implies that each caller must pass down a string and/or format
  string plus arguments

Both challenges can be solved by returning errors:
- the stacktrace is taken at that level where the error is
  treated as a failure instead of passing back an error, i.e.
  inside the It callback
- traditional error wrapping can add additional information, if
  desirable

What was missing was some easy way to generate an error via a gomega
assertion. The new infrastructure achieves that by mirroring the
Gomega/Assertion/AsyncAssertion interfaces with errors as return values instead
of calling a fail handler.

It is intentionally less flexible than the gomega APIs:
- A context must be passed to Eventually/Consistently as first
  parameter because that is needed for proper timeout handling.
- No additional text can be added to the failure through this
  API because error wrapping is meant to be used for this.
- No need to adjust the callstack offset because no backtrace
  is recorded when a failure occurs.

To avoid the useless "unexpected error" log message when passing back a gomega
failure, ExpectNoError gets extended to recognize such errors and then skips
the logging.

pohly commented on test/e2e/framework/expect.go:
```diff
@@ -17,12 +17,203 @@ limitations under the License.
+	if a.interval != nil {
+		assertion = assertion.WithPolling(*a.timeout)
+	}
+	return err, assertion
```
Prow jobs are currently failing, most likely because the default polling interval is very short and thus causes API server overload.

My intention is to get PR #114783 merged, then use the same defaults here as for the global Gomega in that PR, i.e. it will not be necessary to override the polling interval on every usage of this API here.

pohly commented:
Another example of a (currently) poor failure message:
```
{ failed [FAILED] while waiting for hostPath init pod to succeed: error while waiting for pod provisioning-6092/hostpath-symlink-prep-provisioning-6092 to be Succeeded or Failed: pod "hostpath-symlink-prep-provisioning-6092" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.4 PodIP:10.244.2.86 PodIPs:[{IP:10.244.2.86}] StartTime:2023-01-09 20:21:32 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:init-volume-provisioning-6092 State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,StartedAt:2023-01-09 20:21:34 +0000 UTC,FinishedAt:2023-01-09 20:21:34 +0000 UTC,ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:registry.k8s.io/e2e-test-images/busybox:1.29-4 ImageID:registry.k8s.io/e2e-test-images/busybox@sha256:2e0f836850e09b8b7cc937681d6194537a09fbd5f6b9e08f4d646a85128e8937 ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b Started:0xc0046ccddf}] QOSClass:BestEffort EphemeralContainerStatuses:[]}
In [It] at: test/e2e/storage/drivers/in_tree.go:842 @ 01/09/23 20:21:58.367
}
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114917/pull-kubernetes-e2e-kind/1612536279373713408

pohly commented:
And another one:

```
Jan  9 20:21:58.367: INFO: Unexpected error: while waiting for hostPath init pod to succeed: 
    <*fmt.wrapError | 0xc0002738c0>: {
        msg: "error while waiting for pod provisioning-6092/hostpath-symlink-prep-provisioning-6092 to be Succeeded or Failed: pod \"hostpath-symlink-prep-provisioning-6092\" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.4 PodIP:10.244.2.86 PodIPs:[{IP:10.244.2.86}] StartTime:2023-01-09 20:21:32 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:init-volume-provisioning-6092 State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,StartedAt:2023-01-09 20:21:34 +0000 UTC,FinishedAt:2023-01-09 20:21:34 +0000 UTC,ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:registry.k8s.io/e2e-test-images/busybox:1.29-4 ImageID:registry.k8s.io/e2e-test-images/busybox@sha256:2e0f836850e09b8b7cc937681d6194537a09fbd5f6b9e08f4d646a85128e8937 ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b Started:0xc0046ccddf}] QOSClass:BestEffort EphemeralContainerStatuses:[]}",
        err: <*errors.errorString | 0xc000dcc6d0>{
            s: "pod \"hostpath-symlink-prep-provisioning-6092\" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.4 PodIP:10.244.2.86 PodIPs:[{IP:10.244.2.86}] StartTime:2023-01-09 20:21:32 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:init-volume-provisioning-6092 State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,StartedAt:2023-01-09 20:21:34 +0000 UTC,FinishedAt:2023-01-09 20:21:34 +0000 UTC,ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:registry.k8s.io/e2e-test-images/busybox:1.29-4 ImageID:registry.k8s.io/e2e-test-images/busybox@sha256:2e0f836850e09b8b7cc937681d6194537a09fbd5f6b9e08f4d646a85128e8937 ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b Started:0xc0046ccddf}] QOSClass:BestEffort EphemeralContainerStatuses:[]}",
        },
    }
[FAILED] while waiting for hostPath init pod to succeed: error while waiting for pod provisioning-6092/hostpath-symlink-prep-provisioning-6092 to be Succeeded or Failed: pod "hostpath-symlink-prep-provisioning-6092" failed with status: {Phase:Failed Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:ContainersReady Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason:PodFailed Message:} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2023-01-09 20:21:32 +0000 UTC Reason: Message:}] Message: Reason: NominatedNodeName: HostIP:172.18.0.4 PodIP:10.244.2.86 PodIPs:[{IP:10.244.2.86}] StartTime:2023-01-09 20:21:32 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:init-volume-provisioning-6092 State:{Waiting:nil Running:nil Terminated:&ContainerStateTerminated{ExitCode:1,Signal:0,Reason:Error,Message:,StartedAt:2023-01-09 20:21:34 +0000 UTC,FinishedAt:2023-01-09 20:21:34 +0000 UTC,ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b,}} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:false RestartCount:0 Image:registry.k8s.io/e2e-test-images/busybox:1.29-4 ImageID:registry.k8s.io/e2e-test-images/busybox@sha256:2e0f836850e09b8b7cc937681d6194537a09fbd5f6b9e08f4d646a85128e8937 ContainerID:containerd://4ee3befa399e0ea04144aff31f737ee7c5266846ec904df9f0cb8715f4b3a67b Started:0xc0046ccddf}] QOSClass:BestEffort EphemeralContainerStatuses:[]}
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114917/pull-kubernetes-e2e-kind/1612536279373713408 (stderr of the failed test).

claudiubelu commented:
/remove-sig windows

Add us if you need something specific to Windows.

cici37 commented:
/remove-sig api-machinery

pohly committed:
e2e framework: include additional stack backtrace in failures

When a Gomega failure is converted to an error, the stack at the time when the
failure occurs may be useful: error wrapping provides some bread crumbs that
can be followed to determine where the failure really occurred, but error
wrapping may be missing or ambiguous.

To provide the additional information, a FailureError now includes a full stack
backtrace. The backtrace intentionally makes no attempt to exclude framework
functions besides the gomega support itself because helpers like
e2e/framework/pod may be relevant.

That backtrace is not included in the failure message for the sake of
brevity. Instead, it gets logged as part of the test's output.

pohly committed:
e2e: remove unused label filter from WaitForPodsRunningReady

None of the users of the functions passed anything other than nil or an empty
map and the implementation ignore the parameter - it seems like a candidate for
simplification.

pohly committed:
e2e framework: turn function into gomega.Matcher

The intention is to use this inside a helper function where the
corresponding Expect call is known.

pohly committed:
e2e pod: use gomega.Eventually in WaitForPodsRunningReady

The code becomes simpler (78 insertions, 91 deletions), easier to read (all
code entirely inside WaitForPodsRunningReady, no need to declare and later
overwrite variables) and possibly more correct (if all API calls failed,
the resulting error was ignored when allowedNotReadyPods > 0).

pohly committed:
e2e framework: support getting list of objects

This is similar to the previous support for getting a single object.

pohly committed:
e2e pod: use gomega.Eventually in WaitTimeoutForPodReadyInNamespace/WaitForPodCondition

These get converted together because they relied on FinalErr which now isn't
needed anymore.

pohly committed:
e2e: consolidate checking a pod list

WaitForPods is now a generic function which lists pods and then checks the pods
that it found against some provided condition. A parameter determines how many
pods must be found resp. match the condition for the check to succeed.

pohly committed:
e2e: replace WaitForPodToDisappear with WaitForPodNotFoundInNamespace

WaitForPodToDisappear was always called such that it listed all pods, which
made it less efficient than trying to get just the one pod it was checking for.

Being able to customize the poll interval in practice wasn't useful, therefore
it can be replaced with WaitForPodNotFoundInNamespace.

pohly committed:
e2e pod: use gomega.Eventually in WaitForPodNotFoundInNamespace

pohly committed:
e2e: consolidate pod response checking

This renames PodsResponding to WaitForPodsResponding for the sake of
consistency and adds a timeout parameter. That is necessary because some other
users of NewProxyResponseChecker used a much lower timeout (2min vs. 15min).

Besides simplifying some code, it also makes it easier to rewrite
ProxyResponseChecker because it only gets used in WaitForPodsResponding.

pohly committed:
e2e pod: convert ProxyResponseChecker into matcher

Instead of pod responses being printed to the log each time polling fails, we
get a consolidated failure message with all unexpected pod responses if (and
only if) the check times out or a progress report gets produced.

pohly committed:
e2e pod: use gomega.Eventually in WaitForRestartablePods

pohly committed:
e2e pod: remove dead code

pohly committed:
e2e storage: remove WaitForPodTerminatedInNamespace

Calling WaitForPodTerminatedInNamespace after testFlexVolume is useless because
the client pod that it waits for always gets deleted by testVolumeClient:

https://github.com/kubernetes/kubernetes/blob/0fcc3dbd55732c501d4948840f639e00a2fe33c5/test/e2e/framework/volume/fixtures.go#L541-L546

Worse, because WaitForPodTerminatedInNamespace treats "not found" as "must keep
polling", these two tests always kept waiting for 5 minutes:

    Kubernetes e2e suite: [It] [sig-storage] Flexvolumes should be mountable
    when non-attachable 	6m4s

The only reason why these tests passed is that WaitForPodTerminatedInNamespace
used to return the "not found" API error. That is not guaranteed and about to
change.

pohly commented:
/retest

pohly committed:
e2e pod: unit test for pod status + API error

This covers new behavior in gomega.

claudiubelu commented:
/remove-sig windows

pohly committed:
e2e: use error wrapping with %w

The recently introduced failure handling in ExpectNoError depends on error
wrapping: if an error prefix gets added with `fmt.Errorf("foo: %v", err)`, then
ExpectNoError cannot detect that the root cause is an assertion failure and
then will add another useless "unexpected error" prefix and will not dump the
additional failure information (currently the backtrace inside the E2E
framework).

Instead of manually deciding on a case-by-case basis where %w is needed, all
error wrapping was updated automatically with

    sed -i "s/fmt.Errorf\(.*\): '*\(%s\|%v\)'*\",\(.* err)\)/fmt.Errorf\1: %w\",\3/" $(git grep -l 'fmt.Errorf' test/e2e*)

This may be unnecessary in some cases, but it's not wrong.

pohly commented:
@aojea: this is ready for review. I cleaned up the commit history, so each commit can stand on its own.

cici37 commented:
/remove-sig api-machinery

SergeyKanzhelev commented:
@xmcqueen for sig node lgtm

xmcqueen reviewed (changes requested):
I left one comment.  Its probably nothing. I think I'd like you to doublecheck that one thing.  Thanks

pohly commented on test/e2e/storage/flexvolume.go:
```diff
@@ -197,11 +196,6 @@ var _ = utils.SIGDescribe("Flexvolumes", func() {
 
-		ginkgo.By("waiting for flex client pod to terminate")
-		if err := e2epod.WaitForPodTerminatedInNamespace(ctx, f.ClientSet, config.Prefix+"-client", "", f.Namespace.Name); !apierrors.IsNotFound(err) {
-			framework.ExpectNoError(err, "Failed to wait client pod terminated: %v", err)
```
The commit message for this change explains it:

https://github.com/kubernetes/kubernetes/pull/113298/commits/b3366ce8952daae4a69865134e3180825107cebc

> Calling WaitForPodTerminatedInNamespace after testFlexVolume is useless because
> the client pod that it waits for always gets deleted by testVolumeClient:
> 
> https://github.com/kubernetes/kubernetes/blob/0fcc3dbd55732c501d4948840f639e00a2fe33c5/test/e2e/framework/volume/fixtures.go#L541-L546
> 
> Worse, because WaitForPodTerminatedInNamespace treats "not found" as "must keep
> polling", these two tests always kept waiting for 5 minutes:
> 
>     Kubernetes e2e suite: [It] [sig-storage] Flexvolumes should be mountable
>     when non-attachable 	6m4s
> 
> The only reason why these tests passed is that WaitForPodTerminatedInNamespace
> used to return the "not found" API error. That is not guaranteed and about to
> change.

xmcqueen reviewed (approved):
/lgtm

SergeyKanzhelev commented:
/retest

pohly commented:
/hold

Let's hold this over the weekend before merging it, in case that others also have an opinion.

@aojea: last chance to comment :grin:

aojea commented:
> Let's hold this over the weekend before merging it, in case that others also have an opinion.

I don't have time for review this large PR sorry, but I trust you and the reviewers so feel free to unhold.

My advice, the one that I always do :), merge on weekend and let it soak for at least 1 week, and monitor the CI daily too look for possible changes

Thanks

pohly commented:
Okay, let's merge it then today.

/hold cancel

pohly commented:
https://github.com/kubernetes/kubernetes/pull/113298/commits/6eea1b2efa095bfae444849d9a8981307f79dfe8 failed to swap pod name and namespace when replacing WaitForPodToDisappear with WaitForPodNotFoundInNamespace.

Making the API safer to use by using the same order everywhere (horribly messed up right now!) or not putting two strings next to each other at all is out of scope. Let's fix the commit and the merge again:
```diff
diff --git a/test/e2e/framework/network/utils.go b/test/e2e/framework/network/utils.go
index c0023eaf889..e4f733a57e6 100644
--- a/test/e2e/framework/network/utils.go
+++ b/test/e2e/framework/network/utils.go
@@ -888,7 +888,7 @@ func (config *NetworkingTestConfig) DeleteNetProxyPod(ctx context.Context) {
 	framework.ExpectNoError(config.getPodClient().Delete(ctx, pod.Name, *metav1.NewDeleteOptions(0)))
 	config.EndpointPods = config.EndpointPods[1:]
 	// wait for pod being deleted.
-	err := e2epod.WaitForPodNotFoundInNamespace(ctx, config.f.ClientSet, config.Namespace, pod.Name, wait.ForeverTestTimeout)
+	err := e2epod.WaitForPodNotFoundInNamespace(ctx, config.f.ClientSet, pod.Name, config.Namespace, wait.ForeverTestTimeout)
 	if err != nil {
 		framework.Failf("Failed to delete %s pod: %v", pod.Name, err)
 	}
diff --git a/test/e2e/framework/pod/pod_client.go b/test/e2e/framework/pod/pod_client.go
index 8c323d5a3e7..946587ae1df 100644
--- a/test/e2e/framework/pod/pod_client.go
+++ b/test/e2e/framework/pod/pod_client.go
@@ -181,7 +181,7 @@ func (c *PodClient) DeleteSync(ctx context.Context, name string, options metav1.
 	if err != nil && !apierrors.IsNotFound(err) {
 		framework.Failf("Failed to delete pod %q: %v", name, err)
 	}
-	framework.ExpectNoError(WaitForPodNotFoundInNamespace(ctx, c.f.ClientSet, namespace, name, timeout), "wait for pod %q to disappear", name)
+	framework.ExpectNoError(WaitForPodNotFoundInNamespace(ctx, c.f.ClientSet, name, namespace, timeout), "wait for pod %q to disappear", name)
 }
 
 // mungeSpec apply test-suite specific transformations to the pod spec.
diff --git a/test/e2e/framework/volume/fixtures.go b/test/e2e/framework/volume/fixtures.go
index 3f33a5ec123..d8b62b5b0dc 100644
--- a/test/e2e/framework/volume/fixtures.go
+++ b/test/e2e/framework/volume/fixtures.go
@@ -463,7 +463,7 @@ func runVolumeTesterPod(ctx context.Context, client clientset.Interface, timeout
 	}
 	if err != nil {
 		e2epod.DeletePodOrFail(ctx, client, clientPod.Namespace, clientPod.Name)
-		_ = e2epod.WaitForPodNotFoundInNamespace(ctx, client, clientPod.Namespace, clientPod.Name, timeouts.PodDelete)
+		_ = e2epod.WaitForPodNotFoundInNamespace(ctx, client, clientPod.Name, clientPod.Namespace, timeouts.PodDelete)
 		return nil, err
 	}
 	return clientPod, nil
@@ -541,7 +541,7 @@ func testVolumeClient(ctx context.Context, f *framework.Framework, config TestCo
 		// testVolumeClient might get used more than once per test, therefore
 		// we have to clean up before returning.
 		e2epod.DeletePodOrFail(ctx, f.ClientSet, clientPod.Namespace, clientPod.Name)
-		framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, clientPod.Namespace, clientPod.Name, timeouts.PodDelete))
+		framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, clientPod.Name, clientPod.Namespace, timeouts.PodDelete))
 	}()
 
 	testVolumeContent(f, clientPod, "", fsGroup, fsType, tests)
@@ -576,7 +576,7 @@ func InjectContent(ctx context.Context, f *framework.Framework, config TestConfi
 		// This pod must get deleted before the function returns becaue the test relies on
 		// the volume not being in use.
 		e2epod.DeletePodOrFail(ctx, f.ClientSet, injectorPod.Namespace, injectorPod.Name)
-		framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, injectorPod.Namespace, injectorPod.Name, timeouts.PodDelete))
+		framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, injectorPod.Name, injectorPod.Namespace, timeouts.PodDelete))
 	}()
 
 	ginkgo.By("Writing text file contents in the container.")
diff --git a/test/e2e/kubectl/kubectl.go b/test/e2e/kubectl/kubectl.go
index a8c212bda5c..f5574e83cfd 100644
--- a/test/e2e/kubectl/kubectl.go
+++ b/test/e2e/kubectl/kubectl.go
@@ -734,7 +734,7 @@ metadata:
 				if !strings.Contains(ee.String(), "timed out") {
 					framework.Failf("Missing expected 'timed out' error, got: %#v", ee)
 				}
-				framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, ns, "failure-3", 2*v1.DefaultTerminationGracePeriodSeconds*time.Second))
+				framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, "failure-3", ns, 2*v1.DefaultTerminationGracePeriodSeconds*time.Second))
 			})
 
 			ginkgo.It("[Slow] running a failing command with --leave-stdin-open", func(ctx context.Context) {
diff --git a/test/e2e/storage/pd.go b/test/e2e/storage/pd.go
index 755af5a08f1..8aa86b58a2c 100644
--- a/test/e2e/storage/pd.go
+++ b/test/e2e/storage/pd.go
@@ -195,7 +195,7 @@ var _ = utils.SIGDescribe("Pod Disks [Feature:StorageProvider]", func() {
 					ginkgo.By("deleting host0Pod") // delete this pod before creating next pod
 					framework.ExpectNoError(podClient.Delete(ctx, host0Pod.Name, podDelOpt), "Failed to delete host0Pod")
 					framework.Logf("deleted host0Pod %q", host0Pod.Name)
-					e2epod.WaitForPodNotFoundInNamespace(ctx, cs, host0Pod.Namespace, host0Pod.Name, f.Timeouts.PodDelete)
+					e2epod.WaitForPodNotFoundInNamespace(ctx, cs, host0Pod.Name, host0Pod.Namespace, f.Timeouts.PodDelete)
 					framework.Logf("deleted host0Pod %q disappeared", host0Pod.Name)
 				}
 
diff --git a/test/e2e/storage/testsuites/provisioning.go b/test/e2e/storage/testsuites/provisioning.go
index 4a5dc6e151b..87395546363 100644
--- a/test/e2e/storage/testsuites/provisioning.go
+++ b/test/e2e/storage/testsuites/provisioning.go
@@ -845,7 +845,7 @@ func (t StorageClassTest) TestBindingWaitForFirstConsumerMultiPVC(ctx context.Co
 	framework.ExpectNoError(err)
 	ginkgo.DeferCleanup(func(ctx context.Context) error {
 		e2epod.DeletePodOrFail(ctx, t.Client, pod.Namespace, pod.Name)
-		return e2epod.WaitForPodNotFoundInNamespace(ctx, t.Client, pod.Namespace, pod.Name, t.Timeouts.PodDelete)
+		return e2epod.WaitForPodNotFoundInNamespace(ctx, t.Client, pod.Name, pod.Namespace, t.Timeouts.PodDelete)
 	})
 	if expectUnschedulable {
 		// Verify that no claims are provisioned.
diff --git a/test/e2e/storage/ubernetes_lite_volumes.go b/test/e2e/storage/ubernetes_lite_volumes.go
index 8919a4898e4..252c1015502 100644
--- a/test/e2e/storage/ubernetes_lite_volumes.go
+++ b/test/e2e/storage/ubernetes_lite_volumes.go
@@ -100,7 +100,7 @@ func PodsUseStaticPVsOrFail(ctx context.Context, f *framework.Framework, podCoun
 			go func(config *staticPVTestConfig) {
 				defer ginkgo.GinkgoRecover()
 				defer wg.Done()
-				err := e2epod.WaitForPodNotFoundInNamespace(ctx, c, ns, config.pod.Name, f.Timeouts.PodDelete)
+				err := e2epod.WaitForPodNotFoundInNamespace(ctx, c, config.pod.Name, ns, f.Timeouts.PodDelete)
 				framework.ExpectNoError(err, "while waiting for pod to disappear")
 				errs := e2epv.PVPVCCleanup(ctx, c, ns, config.pv, config.pvc)
 				framework.ExpectNoError(utilerrors.NewAggregate(errs), "while cleaning up PVs and PVCs")
diff --git a/test/e2e/windows/density.go b/test/e2e/windows/density.go
index 76854b7da12..e0ab84ae10f 100644
--- a/test/e2e/windows/density.go
+++ b/test/e2e/windows/density.go
@@ -276,7 +276,7 @@ func deletePodsSync(ctx context.Context, f *framework.Framework, pods []*v1.Pod)
 			err := e2epod.NewPodClient(f).Delete(ctx, pod.ObjectMeta.Name, *metav1.NewDeleteOptions(30))
 			framework.ExpectNoError(err)
 
-			err = e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, f.Namespace.Name, pod.ObjectMeta.Name, 10*time.Minute)
+			err = e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, 10*time.Minute)
 			framework.ExpectNoError(err)
 		}(pod)
 	}
diff --git a/test/e2e_node/node_problem_detector_linux.go b/test/e2e_node/node_problem_detector_linux.go
index 904804cac03..39d25e9765c 100644
--- a/test/e2e_node/node_problem_detector_linux.go
+++ b/test/e2e_node/node_problem_detector_linux.go
@@ -433,7 +433,7 @@ current-context: local-context
 			ginkgo.By("Delete the node problem detector")
 			framework.ExpectNoError(e2epod.NewPodClient(f).Delete(ctx, name, *metav1.NewDeleteOptions(0)))
 			ginkgo.By("Wait for the node problem detector to disappear")
-			gomega.Expect(e2epod.WaitForPodNotFoundInNamespace(ctx, c, ns, name, pollTimeout)).To(gomega.Succeed())
+			gomega.Expect(e2epod.WaitForPodNotFoundInNamespace(ctx, c, name, ns, pollTimeout)).To(gomega.Succeed())
 			ginkgo.By("Delete the config map")
 			framework.ExpectNoError(c.CoreV1().ConfigMaps(ns).Delete(ctx, configName, metav1.DeleteOptions{}))
 			ginkgo.By("Clean up the events")
diff --git a/test/e2e_node/resource_collector.go b/test/e2e_node/resource_collector.go
index 5472213511a..60178b9d764 100644
--- a/test/e2e_node/resource_collector.go
+++ b/test/e2e_node/resource_collector.go
@@ -383,7 +383,7 @@ func deletePodsSync(ctx context.Context, f *framework.Framework, pods []*v1.Pod)
 				framework.Failf("Unexpected error trying to delete pod %s: %v", pod.Name, err)
 			}
 
-			framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, f.Namespace.Name, pod.ObjectMeta.Name, 10*time.Minute))
+			framework.ExpectNoError(e2epod.WaitForPodNotFoundInNamespace(ctx, f.ClientSet, pod.ObjectMeta.Name, f.Namespace.Name, 10*time.Minute))
 		}()
 	}
 	wg.Wait()
```

pohly commented:
/reopen

Not sure whether this is possible for a reverted PR...
